### ISSUE TITLE
Support varied input sequence lengths with a fixed block mask

### ISSUE THREAD

**Issue by tilmto** [pytorch-labs/attention-gym#31]
Thanks for the great repo!

When using a custom-defined attention mask pattern (e.g., the A-shape mask in this [work](https://arxiv.org/pdf/2308.16137)), I noticed that when the input length (e.g., 512) is shorter than the length of the predefined block mask (e.g., 1024) in `mask_mod`, the generation results may not be correct, even though the attention pattern of the former is a truncated version of the latter.

Therefore, I wonder if FlexAttention generally supports varying input sequence lengths under a fixed block mask, and how it handles this situation?

**Comment by tilmto** [comment]
Thanks for the prompt response! So, can I understand it like this: if we need to perform evaluations on common LM benchmarks, which often contain questions of varying lengths, we need to create the block mask on the fly for each input (ideally with _compile=True to speed up this process)?

**Comment by tilmto** [comment]
Got it! The last question is that I find that sometimes setting `_compile=True` leads to errors related to insufficient cache sizes. This often happens with models that have many full attentions, but when replacing them with sliding window attentions, everything works well. Are there any workarounds for this?

**Comment by drisspg** [comment]
hmmm this is likely a dynamic shapes thing, @Chillee

**Comment by drisspg** [comment]
We make this explicit now and the blockmask is tied to an input seqlen. If you can gurantee the  visit pattern + mask mod should work with updated seqlens the user is requrired to just set the seqlen on the BM to the updated input seqlen
